File: 53.最大子序和.js
/*
 * @lc app=leetcode.cn id=53 lang=javascript
 *
 * [53] 最大子序和
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxSubArray = function (nums) {
    let res = nums[0];
    let sum = 0;
    for (let i = 0; i < nums.length; i++) {
        const num = nums[i];
        if (sum > 0) {
            sum += num;
        } else {
            sum = num;
        }
        res = Math.max(res, sum);
        // console.log(`${num}\t${sum}\t${res}`);
    }
    // console.log(res);
    return res;
};
// @lc code=end


maxSubArray([-2, 1, -3, 4, -1, 2, 1, -5, 4]);

File: 53.最大子序和.py
#
# @lc app=leetcode.cn id=53 lang=python3
#
# [53] 最大子序和
#

from typing import List
# @lc code=start


class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        n = len(nums)
        if n < 1:
            return 0

        dp = [0] * n
        dp[0] = nums[0]
        max_v = dp[0]
        for i in range(1, n):
            dp[i] = nums[i]
            if dp[i - 1] > 0:
                dp[i] += dp[i-1]
            max_v = max(max_v, dp[i])
        return max_v


# @lc code=end

solution = Solution()
ans = solution.maxSubArray([8, -19, 5, -4, 20])
print(ans)
